### PR TITLE
proper error message for fn source/sink when there are non-KRM resources

### PR DESCRIPTION
**Pull request by natasha41575**
fixes https://github.com/GoogleContainerTools/kpt/issues/2039

sample output: 

```
$ kpt fn source
error: f1.yaml: resource must have `metadata.name`
```

### CODE REVIEW

File: pkg/api/kptfile/v1alpha2/validation.go [hunk 1]
```diff
@@ -186,8 +186,12 @@ func IsKRM(n *yaml.RNode) error {
 
 func AreKRM(nodes []*yaml.RNode) error {
 	for i := range nodes {
+		// don't do this check for .json files
+		path, _, _ := kioutil.GetFileAnnotations(nodes[i])
+		if strings.HasSuffix(path, ".json") {
```
Comment (droot) [inline pkg/api/kptfile/v1alpha2/validation.go:191]: why are we excluding `.json` files ? json can contain KRM resources.. right ?
Is it motivated by the use-case that involves schema files as `.json` ?

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink.go [hunk 2]
```diff
@@ -46,8 +48,18 @@ func (r *SinkRunner) runE(c *cobra.Command, args []string) error {
 	}
 	outputs := []kio.Writer{&kio.LocalPackageWriter{PackagePath: dir}}
 
+	f := kio.FilterFunc(func(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+		if err := kptfile.AreKRM(nodes); err != nil {
```
Comment (droot) [inline thirdparty/cmdconfig/commands/cmdsink/cmdsink.go:52]: at some point, define a filter that can be re-used in all places will be useful, I think.

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go [hunk 3]
```diff
@@ -36,7 +37,8 @@ items:
       config.kubernetes.io/path: 'f1.yaml'
   spec:
     replicas: 1
-- kind: Service
+- apiVersion: apps/v1
```
[the remark this answers is not in the document]
Comment (droot) [reply thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go:40]: That's a good catch. I wonder if `kubeval` will catch such errors if used as a validator by default.

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink.go (hunk 2 above)
Comment (droot) [reply thirdparty/cmdconfig/commands/cmdsink/cmdsink.go:52]: @phanimarupaka is making changes to `sink`, so this change might be affected by his changes.

File: pkg/api/kptfile/v1alpha2/validation.go (hunk 1 above)
Comment (natasha41575) [reply pkg/api/kptfile/v1alpha2/validation.go:191]: `TestSinkCommandJSON` failed without this check. I'm not sure if the readers and writers in kyaml are intended to be used for JSON.

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (natasha41575) [reply thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go:40]: `kubeval` can only catch such errors once openapi schema can be injected into functions, right? or is there another way?

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink.go (hunk 2 above)
Comment (natasha41575) [reply thirdparty/cmdconfig/commands/cmdsink/cmdsink.go:52]: > at some point, define a filter that can be re-used in all places will be useful, I think.

will do

File: thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (natasha41575) [reply thirdparty/cmdconfig/commands/cmdsink/cmdsink_test.go]: Also, changed to v1

File: pkg/api/kptfile/v1alpha2/validation.go (hunk 1 above)
Comment (natasha41575) [reply pkg/api/kptfile/v1alpha2/validation.go:191]: `TestSourceCommandJSON` and `TestSinkCommandJSON` failed without this check. I'm not sure if the readers and writers in kyaml are intended to be used for JSON.

Comment (droot) [reply pkg/api/kptfile/v1alpha2/validation.go:191]: resources can exist in `json` and `yaml`, so we need to be enforcing on both. But I remember cases where people are using `schema` files in packages and they will be non `non-KRM`. 

I have seen `source` and `sink` being used to stage or prep-up the kpt packages, so these might deal with non-KRM resources, so I am thinking if we should be enforcing KRM checks in `source` and `sink`. sort of keeping them dumb.

Comment (frankfarzan) [reply pkg/api/kptfile/v1alpha2/validation.go:191]: Can you show the use case using concrete commands?

Note on JSON vs YAML: Do we actually have e2e tests that use JSON. I'm not sure if anyone uses JSON manifests, but we shouldn't claim it works unless it's tested.

Comment (natasha41575) [reply pkg/api/kptfile/v1alpha2/validation.go:191]: > But I remember cases where people are using schema files in packages and they will be non non-KRM.

If we enforce on json, the use case described in https://github.com/GoogleContainerTools/kpt/issues/2144 will break (they are using a JSON OpenAPI schema file as input to kustomize)

**Comment by frankfarzan** [comment]
What's the behavior when encountering non-KRM for other commands such as `render`, `eval`, etc.? Are they consistent?

**Comment by natasha41575** [comment]
> What's the behavior when encountering non-KRM for other commands such as render, eval, etc.? Are they consistent? If not, what's the plan for them?
Currently render and eval currently throw an error when encountering a non-KRM resource. Not sure what the plan is if we want to make it consistent with source/sink. @droot ?

**Comment by phanimarupaka** [comment]
@natasha41575 What is the status of this PR ?

**Comment by droot** [comment]
@natasha41575 @phanimarupaka It is safe to close this PR because we don't need to enforce the invariants that we enforce in `eval` and `render` in `source` etc because if `source` is used with `eval` and `render`, then the invariants will be enforced anyways. And in all other cases where `source` is used without `eval` or `render`, we are not sure about enforcing the invariants. We can pick this later if we run in to cases.

closing this for now.

**Pull request closed without merging**